File: Mastermind.py
from array import array
import random

sequence = array('c', [' ', ' ', ' ', ' '])
turnsTaken = 1
combinationNotYetFound = True


def generate_random_number():
    return random.randint(0, 4)


def get_letter_from_number(number):
    switcher = {
        0: 'R',
        1: 'G',
        2: 'B',
        3: 'Y'
    }
    return switcher.get(number, 'R')


def generate_combination():
    global sequence
    for x in range(0, 4):
        sequence[x]= get_letter_from_number(generate_random_number())


def check_against_solution(guess):
    matches = 0
    fixed = guess.upper()
    for x in range(0,4):
        if fixed[x] == sequence[x]:
            matches += 1
    print("There were %d matches." % matches)
    return matches


def request_guess():
    global turnsTaken
    guess = ""
    while len(guess) != 4:
        guess = raw_input("[Turn # %d]\tPlease enter a guess (XXXX from RGBY): " % turnsTaken)
        if len(guess) != 4:
            print("ERROR: Supplied guess is in incorrect format!")
        else:
            turnsTaken += 1
            progress = check_against_solution(guess)
            if progress == 4:
                print("You found the correct combination!")
                global combinationNotYetFound
                combinationNotYetFound = False


def game():
    generate_combination()
    while combinationNotYetFound:
        request_guess()


def main():
    game()


if __name__== "__main__":
    main()

File: Mastermind.java
package MastermindGame;

import java.util.Scanner;

/**
 * This class is a version of Mastermind, a classic game of
 * combination deduction.
 *
 * @author malloneec
 * @since 8/8/16
 */
public class Mastermind {
    public static void main(String[] args){
        Game myGame = new Game();
        myGame.begin();
    }

    private static class Game {
        private char[] sequence;
        private int turnsTaken;
        private Scanner reader = new Scanner(System.in);
        boolean correctCombinationNotFound;

        private Game(){
            turnsTaken = 1;
            sequence = new char[4];
            correctCombinationNotFound = true;
            generateCombination();
        }

        private void generateCombination(){
            for(int i=0 ; i<4; i++){
                int color = (int)(Math.random()*4.0);
                switch(color){
                    case 0:
                        sequence[i] = 'R';
                        break;
                    case 1:
                        sequence[i] = 'G';
                        break;
                    case 2:
                        sequence[i] = 'B';
                        break;
                    case 3:
                        sequence[i] = 'Y';
                        break;
                    default:
                        sequence[i] = 'R';
                }
            }
        }

        public void begin(){
            while(correctCombinationNotFound){
                requestGuess();
            }
        }

        /**
         * No error-checking for correct "colors".
         */
        private void requestGuess(){
            String guess = "";
            while(guess.length()!=4){
                System.out.println("[Turn# "+ turnsTaken + "]\tPlease enter a guess (XXXX from RGBY): ");
                guess = reader.next();
                if(guess.length()!=4){
                    System.out.println("ERROR: Supplied guess is in incorrect format!");
                }else{
                    checkAgainstSolution(guess);
                }
            }
            turnsTaken++;

        }

        private void checkAgainstSolution(String guess){
            String fixed = guess.toUpperCase();
            int matches = 0;
            int misses = 0;
            for(int i = 0; i<4; i++){
                if(fixed.charAt(i)==sequence[i]){
                    matches++;
                }else{
                    for(int j = 0; j<=i; j++) {
                        if(sequence[j]==fixed.charAt(i) && j!=i) {
                            misses++;
                        }
                    }
                }
            }
            if(matches==4){
                correctCombinationNotFound = false;
                System.out.println("You found the correct combination on turn " + turnsTaken + "!!! Congratulations!");
            }else {
                System.out.println("There were " + matches + " matching positions, and " + misses + " misses.");
            }
        }
    }
}
